File: packages/cpp/examples/basic_crud.cpp
/**
 * DBX C++ Example - Basic CRUD Operations
 */

#include <iostream>
#include "dbx.hpp"

using namespace dbx;

int main() {
    try {
        std::cout << "Opening in-memory database..." << std::endl;
        auto db = Database::openInMemory();
        
        // Or open a file-based database:
        // Database db("my_database.db");
        
        // Insert some data
        std::cout << "\nInserting data..." << std::endl;
        db.insert("users", "user:1", "Alice");
        db.insert("users", "user:2", "Bob");
        db.insert("users", "user:3", "Charlie");
        
        // Get data
        std::cout << "\nRetrieving data..." << std::endl;
        if (auto value = db.getString("users", "user:1")) {
            std::cout << "user:1 = " << *value << std::endl;
        }
        
        if (auto value = db.getString("users", "user:2")) {
            std::cout << "user:2 = " << *value << std::endl;
        }
        
        // Count rows
        size_t count = db.count("users");
        std::cout << "\nTotal users: " << count << std::endl;
        
        // Delete a row
        std::cout << "\nDeleting user:2..." << std::endl;
        db.remove("users", "user:2");
        
        // Verify deletion
        if (!db.getString("users", "user:2")) {
            std::cout << "user:2 successfully deleted" << std::endl;
        }
        
        // Count again
        count = db.count("users");
        std::cout << "Total users after deletion: " << count << std::endl;
        
        // Flush to disk
        std::cout << "\nFlushing to disk..." << std::endl;
        db.flush();
        
        std::cout << "\n✓ All operations completed successfully!" << std::endl;
        
        // Database is automatically closed by destructor
        
    } catch (const DatabaseError& e) {
        std::cerr << "Database error: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
}

File: packages/python/examples/basic_crud.py
"""
DBX Python Bindings - Basic CRUD Example

This example demonstrates basic database operations using DBX Python bindings.
"""

from dbx_py import Database


def main():
    # Open an in-memory database
    print("Opening in-memory database...")
    db = Database.open_in_memory()
    
    # Or open a file-based database:
    # db = Database("my_database.db")
    
    try:
        # Insert some data
        print("\nInserting data...")
        db.insert("users", b"user:1", b"Alice")
        db.insert("users", b"user:2", b"Bob")
        db.insert("users", b"user:3", b"Charlie")
        
        # Get data
        print("\nRetrieving data...")
        value = db.get("users", b"user:1")
        if value:
            print(f"user:1 = {value.decode('utf-8')}")
        
        value = db.get("users", b"user:2")
        if value:
            print(f"user:2 = {value.decode('utf-8')}")
        
        # Count rows
        count = db.count("users")
        print(f"\nTotal users: {count}")
        
        # Delete a row
        print("\nDeleting user:2...")
        db.delete("users", b"user:2")
        
        # Verify deletion
        value = db.get("users", b"user:2")
        if value is None:
            print("user:2 successfully deleted")
        
        # Count again
        count = db.count("users")
        print(f"Total users after deletion: {count}")
        
        # Flush to disk (if using file-based database)
        print("\nFlushing to disk...")
        db.flush()
        
        print("\n✓ All operations completed successfully!")
        
    finally:
        # Close the database
        print("\nClosing database...")
        db.close()


if __name__ == "__main__":
    main()
